### ISSUE TITLE
Setting up LDAP authentication

### ISSUE THREAD

**Issue by euanmacinnes** [sqlpad/sqlpad#898]
Hi,
   I've enabled the LDAP authentication using environment variables, in the latest docker image (as of 21 Nov 2020), and can't get the LDAP function to work. I don't see any debug information in the docker logs that says that the LDAP is working or not, only one line that states:

```
"Enabling basic authentication strategy."
"Enabling ldap authentication strategy."
"Enabling local authentication strategy."
```

However, upon attempting to log in a (known valid) user's details, there is just an "invalid username / password" message appearing in a pink box on the site, and no additional output in the debug file (i.e. did it find the LDAP server Ok, did it log in Ok, did it find the name, etc..).

Any help would be appreciated. I've configured both PAM authentication with LDAP, and other apps succesfully, using the following settings:

```
          SQLPAD_LDAP_AUTH_ENABLED: 'true'
          SQLPAD_LDAP_URL: 'ldap://192.168.0.101:389'
          SQLPAD_LDAP_SEARCH_BASE: 'ou=Users,dc=amt,dc=com'
          SQLPAD_LDAP_BIND_DN: 'cn=admin,dc=amt,dc=com'
          SQLPAD_LDAP_PASSWORD: '****************'
          SQLPAD_LDAP_SEARCH_FILTER: '(uid={{username}})'
```

This is connecting to an OpenLDAP server that is on the same private LAN. TLS is not enabled on the LDAP.

**Comment by eladeyal-intel** [comment]
1. Don't forget that for LDAP to kick in, the username provided in the login dialog-box must be without "@"

2. I wonder if re-implementing https://github.com/sqlpad/sqlpad/pull/757 again will add more logs. (The LDAP authorization provider was changed after this PR was merged). Unfortunately I'm not sure I am able to test this now myself.

**Comment by rickbergfalk** [comment]
@eladeyal-intel yeah the new provider probably doesn't have the same logging functionality.

@euanmacinnes are your ldap usernames email addresses?

Is there a better way to handle auth if both local and ldap are used? Should there be a separate button? A separate form? Should  both auth mechanisms be tried?

At the very least perhaps if ldap is enabled and local auth is disabled, the input is always sent to the ldap auth strategy regardless of its content.

**Comment by euanmacinnes** [comment]
> @eladeyal-intel yeah the new provider probably doesn't have the same logging functionality.
> 
> @euanmacinnes are your ldap usernames email addresses?
> 
> Is there a better way to handle auth if both local and ldap are used? Should there be a separate button? A separate form? Should both auth mechanisms be tried?
> 
> At the very least perhaps if ldap is enabled and local auth is disabled, the input is always sent to the ldap auth strategy regardless of its content.

My LDAP names are not email addresses, they're "standard" Linux logins, like "euanmacinnes" for example. No special characters etc..

Typically for the other apps we are using, the LDAP is tried first, if that fails, local is the fallback, so that the LDAP always acts as the "master list", and we only create local accounts for the "hidden godmode" users. This is then specified in the config file. e.g. the order they are listed determines what gets tried first, as one way. Or a priority marker as another (priority 1 or 2, whichever is lower, is first).

The main issue right now is I have no insight into why the LDAP search is failing, so I can't see what SQLPad doesn't like about the configuration.

**Comment by eladeyal-intel** [comment]
@euanmacinnes, take a look at https://github.com/sqlpad/sqlpad/pull/907

With the following configuration parameters I was able to log-in as "tesla" (password "password') and see logging such as
```
 sending request {"messageID":3,"protocolOp":"SearchRequest","baseObject":{"rdns":[{"attrs":{"dc":{"value":"example","name":"dc","order":0}},"spLead":0,"spTrail":0},{"attrs":{"dc":{"value":"com","name":"dc","order":0}},"spLead":0,"spTrail":0}],"_format":{}},"scope":"sub","derefAliases":0,"sizeLimit":0,"timeLimit":10,"typesOnly":false,"filter":"(cn=uid=tesla,dc=example,dc=com)","attributes":[],"controls":[]}                             
```

```
SQLPAD_APP_LOG_LEVEL = trace
SQLPAD_LDAP_DEFAULT_ROLE = "admin"
SQLPAD_USERPASS_AUTH_DISABLED=true
SQLPAD_LDAP_AUTH_ENABLED=true
SQLPAD_LDAP_AUTO_SIGN_UP=true
SQLPAD_LDAP_URL=ldap://ldap.forumsys.com:389
SQLPAD_LDAP_SEARCH_BASE=dc=example,dc=com
SQLPAD_LDAP_SEARCH_FILTER="(uid={{username}})"
SQLPAD_LDAP_BIND_DN=
SQLPAD_LDAP_PASSWORD=
```

**Comment by rickbergfalk** [comment]
Just tagged 6.0.2 which will have your logging update @eladeyal-intel.

I also added a bind check at startup that will print a warning if connection fails.

**Comment by euanmacinnes** [comment]
Ok great, thanks! I'll take a look and update with the result.

**Comment by euanmacinnes** [comment]
I'm getting a really strange error in the docker image:

{"level":50,"time":"2020-11-30T02:17:21.157Z","pid":1,"hostname":"4b1bf772aa54","name":"sqlpad-app","stack":"TypeError: Cannot read property 'toLowerCase' of undefined\n    at Strategy.passportLdapStrategyHandler [as verify] (/usr/app/auth-strategies/ldap.js:102:25)\n    at Strategy.<anonymous> (/usr/app/node_modules/passport-ldapauth/lib/passport-ldapauth/strategy.js:314:21)\n    at /usr/app/node_modules/ldapauth-fork/lib/ldapauth.js:430:18\n    at /usr/app/node_modules/ldapauth-fork/lib/ldapauth.js:373:5\n    at CorkedEmitter.<anonymous> (/usr/app/node_modules/ldapauth-fork/lib/ldapauth.js:281:16)\n    at CorkedEmitter.emit (events.js:315:20)\n    at CorkedEmitter.emit (/usr/app/node_modules/ldapjs/lib/corked_emitter.js:44:33)\n    at sendResult (/usr/app/node_modules/ldapjs/lib/client/client.js:1190:22)\n    at messageCallback (/usr/app/node_modules/ldapjs/lib/client/client.js:1218:16)\n    at Parser.onMessage (/usr/app/node_modules/ldapjs/lib/client/client.js:884:14)\n    at Parser.emit (events.js:315:20)\n    at Parser.write (/usr/app/node_modules/ldapjs/lib/messages/parser.js:107:8)\n    at Socket.onData (/usr/app/node_modules/ldapjs/lib/client/client.js:871:22)\n    at Socket.emit (events.js:315:20)\n    at addChunk (_stream_readable.js:295:12)\n    at readableAddChunk (_stream_readable.js:271:9)","type":"Error","msg":"Cannot read property 'toLowerCase' of undefined"}

**Comment by rickbergfalk** [comment]
@euanmacinnes I'll check this out and push a fix quick

**Comment by rickbergfalk** [comment]
@euanmacinnes the auth handler makes the assumption "mail" profile field will be present for email. I'm not quite sure what to do about a fix as email is a required SQLPad user field. 

I think what ultimately needs to happen is SQLPad's `user.email` needs to be nullable and a separate id column needs to be added. I don't think that's a fix I'll be able to make quickly.

On side note - it looks like the logs you pasted contained some sensitive info. May want to remove the info and handle accordingly.

**Comment by euanmacinnes** [comment]
Ok thanks, I'm able to add the Email attribute to the user accounts, so I
can work around it for now, thanks!

I'll update the log details.

**Comment by euanmacinnes** [comment]
> @euanmacinnes the auth handler makes the assumption "mail" profile field will be present for email. I'm not quite sure what to do about a fix as email is a required SQLPad user field.
> 
@rickbergfalk perhaps a temporary workaround would be to default the email to be the login account name if it's null?

**Comment by rickbergfalk** [comment]
@euanmacinnes I was thinking about that. It doesn't need to be formatted as an email and the app doesn't send email either.

`users.email` serves as an external identifier field so it might be okay to put the account login.

**Comment by rickbergfalk** [comment]
@euanmacinnes curious as to why your LDAP setup is missing the mail attribute. Was this just for testing purposes or was it intended for mail not to be set?

**Comment by euanmacinnes** [comment]
It's intended for mail not to be set, primarily for corporate security
reasons.

**Comment by rickbergfalk** [comment]
I just merged a change to how LDAP will be working for you @euanmacinnes. I've decided to add an `ldap_id` column to SQLPad's `users` table to store profile uid/sAMAccountName depending on LDAP platform used. SQLPad's users `email` column is now nullable, and `ldap_id` will be used as a secondary lookup in the event email is missing.

For now the LDAP login will also check for uid/sAMAccountName in the `users.email` column when trying to find the user in SQLPad's database, so there shouldn't be an impact... but just in case you see any wonky behavior it could be behind these changes. 

I've updated LDAP tests to cover what I could, but I didn't go as far as to stand up an LDAP server with missing mail attributes.
